// prefix-infix-postfix/src/index.js
/**
  The MIT License (MIT) Copyright (c) 2017 MonikaIlieva

  Permission is hereby granted, free of charge, to any person obtaining a copy
  of this software and associated documentation files (the "Software"), to deal
  in the Software without restriction, including without limitation the rights
  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
  copies of the Software, and to permit persons to whom the Software is
  furnished to do so, subject to the following conditions:

  The above copyright notice and this permission notice shall be included in
  all copies or substantial portions of the Software.

  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
  THE SOFTWARE.
*/

(function(root, factory) {
  if(typeof define === "function" && define.amd)
    define(factory);
  else if(typeof module === "object" && module.exports)
    module.exports = factory();
  else
    root['PIP'] = factory();
}(this, function() {
  'use strict';

  function normalize(expr) {
    return expr
        .match(/(?:[0-9a-zA-Z]+|\+|\*|\-|\/|\(|\)|&&|<=>|\|\||==|=>|<=|!=|\!|~|>|<|\^)/g)
        .join(" ");
  }

  function precedence(op) {
    const operatorsPrecedence = {
      '||': 1,
      '&&': 2,
      '<': 3,
      '>': 3,
      '<=': 3,
      '>=': 3,
      '+': 4,
      '-': 4,
      '*': 5,
      '\\': 5,
      '/': 5,
      '^': 6,
    }

  return operatorsPrecedence[op];
}

  const OPERATORS = ['+', '-', '*', '\\', '/', '&&', '||', '<=>', '=>', '~', '!', '(', ')'];

  var Expression = function(expr) {
    if(! (this instanceof Expression))
      return new Expression(expr);

    this.__expr = expr;
    return this;
  }

  Expression.prototype = {
    toPrefix: function(expr) {
      if(expr)
        this.__expr = expr;

      return this.toPostfix(this.__expr.split("").reverse(" ").map(function(value) {
        if(value === '(')
          return ')';
        else if(value === ')')
          return '('
        return value;
      }).join(" ")).split(" ").reverse().join(" ");
    },
    toInfix: function(expr) {
      if(expr !== undefined)
        this.__expr = expr;

      //simple way
      const prefix = this.__expr.replace(/\(/g, "").match(/^(\+|-|\*|\\)+?/);
      const postfix = this.__expr.replace(/\)/g, "").match(/(\+|-|\*|\\)+?$/);

      if(!prefix && !postfix)
        return expr;

      expr = normalize(this.__expr)
             .split(" ")
             .filter(function(value) {
               return !value.match(/\(|\)/g);
             });

      if(prefix)
        expr = expr.reverse();

      var stack = [];
      for(var value of expr) {
        if(value.match(/\+|\*|\-|\/|\(|\)|&&|<=>|\|\||==|=>|<=|!=|\!|~|>|<|\^/g)) {
          var fop = stack.shift();
          var sop = stack.shift();
          var op = value;
          stack.unshift(postfix ? '(' + sop + ' ' + op + ' ' + fop + ')' : 
                                  '(' + fop + ' ' + op + ' ' + sop + ')');
        } else
          stack.unshift(value);
      }

      return stack.pop();
    },
    toPostfix: function(expr) {
      if(expr)
        this.__expr = expr;

      expr = normalize(this.toInfix(this.__expr))
             .split(" ");

      var stack = [];
      var result = "";
      for(var token of expr) {
        if(token === '(')
          stack.push(token);
        else if(token === ')') {
          while(stack[stack.length - 1] !== '(')
            result += stack.pop() + " ";

          if(stack[stack.length - 1] !== '(')
            throw 'Invalid expression';

          stack.pop();
        } else if(OPERATORS.indexOf(token) > -1) {
          if(stack.length) {
            while(precedence(token) <= precedence(stack[stack.length - 1]))
              result += stack.pop() + " ";
          }
          stack.push(token);
        }
        else
          result += token + " ";
      }

      while(stack.length) {
        result += stack.pop() + " ";
      }

      return result.trim();
    }
  }

  Expression.VERSION = '0.0.1';
  return Expression;
}))
